File: Medium/Javascript/1750. Minimum Length of String After Deleting Similar Ends.js
/**Given a string s consisting only of characters 'a', 'b', and 'c'. You are asked to apply the following algorithm on the string any number of times:

Pick a non-empty prefix from the string s where all the characters in the prefix are equal.
Pick a non-empty suffix from the string s where all the characters in this suffix are equal.
The prefix and the suffix should not intersect at any index.
The characters from the prefix and suffix must be the same.
Delete both the prefix and the suffix.
Return the minimum length of s after performing the above operation any number of times (possibly zero times).
 */

/**
 * @param {string} s
 * @return {number}
 */
var minimumLength = function (s) {
  let l = 0;
  let r = s.length - 1;

  while (l < r && s.charAt(l) === s.charAt(r)) {
    let char = s.charAt(l);

    while (l <= r && s.charAt(l) === char) {
      l++;
    }

    while (r <= r && s.charAt(r) === char) {
      r--;
    }
  }

  return l > r ? 0 : r - l + 1;
};

File: Medium/python/1750. Minimum Length of String After Deleting Similar Ends.py
'''
Given a string s consisting only of characters 'a', 'b', and 'c'. You are asked to apply the following algorithm on the string any number of times:

Pick a non-empty prefix from the string s where all the characters in the prefix are equal.
Pick a non-empty suffix from the string s where all the characters in this suffix are equal.
The prefix and the suffix should not intersect at any index.
The characters from the prefix and suffix must be the same.
Delete both the prefix and the suffix.
Return the minimum length of s after performing the above operation any number of times (possibly zero times).
'''


class Solution:
    def minimumLength(self, s: str) -> int:
        l = 0
        r = len(s) - 1

        while l < r and s[l] == s[r]:
            c = s[l]

            while l <= r and s[l] == c:
                l += 1

            while r >= l and s[r] == c:
                r -= 1

        return 0 if l > r else r - l + 1

print(Solution().minimumLength("aca"))
